### PR TITLE
[LLVMGPU][ROCm] Add validation on finalized llvm bitcode

### PR DESCRIPTION
**Pull request by kuhar**
Check that there are no unresolved external functions that will otherwise compile fine but be rejected by the driver.

The validation happens on the llvm bitcode after bitcode linking, when there is no chance for anything to resolve these external functions.

This is to guard against future issues similar to
https://github.com/iree-org/iree/issues/18534 that are close to undebuggable for end users.

### CODE REVIEW

File: compiler/plugins/target/ROCM/test/external_function_validation.mlir [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      hal.executable.export public @test_dispatch_0_elementwise_D_i32 ordinal(0) layout(#hal.pipeline.layout<constants = 2, bindings = [#hal.pipeline.binding<storage_buffer, "ReadOnly|Indirect">, #hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) {
+      ^bb0(%arg0: !hal.device, %arg1: index):
+        %x, %y, %z = flow.dispatch.workgroup_count_from_slice %arg1
+        hal.return %x, %y, %z : index, index, index
+      }
+      builtin.module {
+        func.func private @external_func(i32) -> i32
+
+        func.func @test_dispatch_0_elementwise_D_i32() attributes {translation_info = #iree_codegen.translation_info<LLVMGPUVectorize workgroup_size = [64, 1, 1] subgroup_size = 32>} {
```
Comment (kuhar) [inline compiler/plugins/target/ROCM/test/external_function_validation.mlir]: I think it'd be better to use 'None' as the pipeline, so that this test doesn't depend on how we codegen

File: compiler/plugins/target/ROCM/test/external_function_validation.mlir [hunk 2]
```diff
@@ -0,0 +1,60 @@
+// RUN: iree-opt --pass-pipeline='builtin.module(iree-hal-transformation-pipeline{start=executable-configurations})' \
+// RUN:   --verify-diagnostics %s -o -
+
+// The final bitcode validation should error out on any external functions that
+// remain in the final bitcode (post device bitcode linking).
+
+module attributes {stream.affinity.default = #hal.device.affinity<@__device_0>} {
+  util.global private @__device_0 = #hal.device.target<"hip", {legacy_sync}, [#hal.executable.target<"rocm", "rocm-hsaco-fb", {iree.gpu.target = #iree_gpu.target<arch = "gfx942", features = "", wgp = <compute =  fp64|fp32|fp16|int64|int32|int16|int8, storage =  b64|b32|b16|b8, subgroup =  shuffle|arithmetic, dot =  dp4xi8toi32, mma = [<MFMA_F32_16x16x4_F32>, <MFMA_F32_16x16x16_F16>, <MFMA_F32_32x32x8_F16>, <MFMA_F32_16x16x32_F8E4M3FNUZ>, <MFMA_I32_16x16x32_I8>, <MFMA_I32_32x32x16_I8>], subgroup_size_choices = [64], max_workgroup_sizes = [1024, 1024, 1024], max_thread_count_per_workgroup = 1024, max_workgroup_memory_bytes = 65536, max_workgroup_counts = [2147483647, 2147483647, 2147483647]>>, ukernels = "none"}>]> : !hal.device
```
Comment (qedawkins) [inline compiler/plugins/target/ROCM/test/external_function_validation.mlir]: nit: Can you simplify the target arch in this test? I don't think `dot` and `mma` are needed for the test here (ditto for the variant below).

File: compiler/plugins/target/ROCM/test/external_function_validation.mlir [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    %dense_row_major = hal.encoding_type<dense_row_major> : i32
+    hal.buffer_view.assert<%arg0 : !hal.buffer_view> message("input0") shape([%c4096, %c4096]) type(%element_type_f16) encoding(%dense_row_major)
+    %0 = stream.tensor.import on(#hal.device.affinity<@__device_0>) %arg0 : !hal.buffer_view -> tensor<4096x4096xf16> in !stream.resource<external>{%c33554432}
+    hal.buffer_view.assert<%arg1 : !hal.buffer_view> message("input1") shape([%c4096, %c4096]) type(%element_type_f16) encoding(%dense_row_major)
+    %1 = stream.tensor.import on(#hal.device.affinity<@__device_0>) %arg1 : !hal.buffer_view -> tensor<4096x4096xf16> in !stream.resource<external>{%c33554432}
+    %result, %result_timepoint = stream.resource.alloca uninitialized on(#hal.device.affinity<@__device_0>) : !stream.resource<external>{%c33554432} => !stream.timepoint
+    %2 = stream.cmd.execute on(#hal.device.affinity<@__device_0>) await(%result_timepoint) => with(%0 as %arg2: !stream.resource<external>{%c33554432}, %1 as %arg3: !stream.resource<external>{%c33554432}, %result as %arg4: !stream.resource<external>{%c33554432}) {
+      stream.cmd.dispatch @test::@rocm_hsaco_fb::@test {
```
Comment (qedawkins) [inline compiler/plugins/target/ROCM/test/external_function_validation.mlir]: Do you need this dispatch so that the executable doesn't get CSE'd? Can we just isolate the executable and call that pass for serializing executables? Why do we need to invoke the whole pipeline here?

File: compiler/plugins/target/ROCM/test/external_function_validation.mlir [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+          %c0 = arith.constant 0 : index
+          %thread_id_x = gpu.thread_id  x
+          %thread_id_y = gpu.thread_id  y
+          %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, "ReadOnly|Indirect">, #hal.pipeline.binding<storage_buffer, "ReadOnly|Indirect">, #hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags("ReadOnly|Indirect") : memref<4096x4096xf16, strided<[4096, 1], offset: ?>>
+          %1 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, "ReadOnly|Indirect">, #hal.pipeline.binding<storage_buffer, "ReadOnly|Indirect">, #hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(1) alignment(64) offset(%c0) flags("ReadOnly|Indirect") : memref<4096x4096xf16, strided<[4096, 1], offset: ?>>
+          %2 = vector.load %0[%thread_id_x, %thread_id_y] : memref<4096x4096xf16, strided<[4096, 1], offset: ?>>, vector<1xf16>
+          %3 = func.call @external_func(%2) : (vector<1xf16>) -> vector<1xf16>
```
Comment (qedawkins) [inline compiler/plugins/target/ROCM/test/external_function_validation.mlir]: nit: I think a single ReadWrite binding + func.call on that memref should be good enough, no need to include things like thread ids and vector ops.

**Review by qedawkins — COMMENTED** [review]
This looks like nice additional validation, thanks! Some questions about simplifying the test

File: compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp [hunk 5]
```diff
@@ -39,6 +40,36 @@ struct TransformOptions : public PassPipelineOptions<TransformOptions> {
       llvm::cl::desc("Whether to link hal.executable ops together."),
       llvm::cl::init(true),
   };
+  Option<PipelinePhase> start{
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp]: I'm not super sure we want to expose this - since I don't think we need it here we should drop it or move it to a separate PR to talk about

File: compiler/plugins/target/ROCM/test/external_function_validation.mlir [hunk 6]
```diff
@@ -0,0 +1,60 @@
+// RUN: iree-opt --pass-pipeline='builtin.module(iree-hal-transformation-pipeline{start=executable-configurations})' \
```
Comment (benvanik) [inline compiler/plugins/target/ROCM/test/external_function_validation.mlir]: you can use `--iree-hal-transformation-pipeline` (shouldn't have to specify a start/end) once simplifying this to just be the executable - you can change then to a public `stream.executable` - see `compiler/plugins/target/ROCM/test/smoketest.mlir`

**Review by benvanik — CHANGES_REQUESTED** [review]

File: compiler/plugins/target/ROCM/test/external_function_validation.mlir (hunk 6 above)
Comment (kuhar) [reply compiler/plugins/target/ROCM/test/external_function_validation.mlir]: I used the serialization pipeline instead

**Review by qedawkins — APPROVED** [review]
nice

**Review by benvanik — APPROVED** [review]
